File: typescript/day_2/main.js
const fs_2 = require('fs');
const input_puller_2 = (file_name) => {
    const input = fs_2.readFileSync(file_name, 'utf-8');
    const code_string = input.split(',');
    let code = code_string.map((val) => parseInt(val, 10));
    return code;
};
const run_opcode = (code, instruction_pointer) => {
    let temp_code = code;
    const opcode = temp_code[instruction_pointer];
    const param_1 = temp_code[instruction_pointer + 1];
    const param_2 = temp_code[instruction_pointer + 2];
    const param_3 = temp_code[instruction_pointer + 3];
    if (opcode == 1) {
        temp_code[param_3] = temp_code[param_1] + temp_code[param_2];
    }
    else if (opcode == 2) {
        temp_code[param_3] = temp_code[param_1] * temp_code[param_2];
    }
    return temp_code;
};
const run_program = (code) => {
    let temp_code = code;
    for (let i = 0; i < temp_code.length; i += 4) {
        if (temp_code[i] == 99) {
            break;
            return temp_code;
        }
        temp_code = run_opcode(temp_code, i);
    }
    return temp_code;
};
const get_solution_2 = () => {
    const code = input_puller_2('./day_2/input.txt');
    let temp_code;
    for (let noun = 0; noun < 100; noun++) {
        for (let verb = 0; verb < 100; verb++) {
            temp_code = Array.from(code);
            temp_code[1] = noun;
            temp_code[2] = verb;
            temp_code = run_program(temp_code);
            if (temp_code[0] == 19690720) {
                console.log(`noun = ${temp_code[1]}, verb = ${temp_code[2]}`);
                console.group(`(100 * noun) + verb = ${100 * temp_code[1] + temp_code[2]}`);
                return;
            }
        }
    }
    for (let verb = 0; verb < 100; verb++) {
        for (let noun = 0; noun < 100; noun++) {
            temp_code = Array.from(code);
            temp_code[1] = noun;
            temp_code[2] = verb;
            temp_code = run_program(temp_code);
            if (temp_code[0] == 19690720) {
                console.log(`noun = ${temp_code[1]}, verb = ${temp_code[2]}`);
                return;
            }
        }
    }
    console.log('Fail');
};
get_solution_2();

File: python/day_2/main.py
'''
day_2
'''
from typing import List


def main():
    code = input_puller('input.txt')
    code[1] = 12
    code[2] = 2
    answer = get_solution(code.copy())

    print(answer)
# END main()


def input_puller(file_name: str) -> List[int]:
    code = []

    with open(file_name, 'r', encoding='utf-8') as file:
        code = file.read().strip().split(',')
        code = [int(num) for num in code]
    # END WITH

    return code
# END input_puller()


def run_opcode(code: List[int], instruction_pointer: int):
    opcode = code[instruction_pointer]
    param_1 = code[instruction_pointer + 1]
    param_2 = code[instruction_pointer + 2]
    param_3 = code[instruction_pointer + 3]

    if opcode == 1:
        code[param_3] = code[param_1] + code[param_2]
    elif opcode == 2:
        code[param_3] = code[param_1] * code[param_2]
    # END IF
# END run_opcode()


def run_program(code: List[int]):
    for i in range(0, len(code), 4):
        if code[i] == 99:
            break
        # END IF

        run_opcode(code, i)
    # END FOR
# END run_program()


def get_solution(code: List[int]) -> int:
    temp_code: List[int]

    for noun in range(0, 100):
        for verb in range(0, 100):
            temp_code = code.copy()
            temp_code[1] = noun
            temp_code[2] = verb
            run_program(temp_code)

            if (temp_code[0] == 19690720):
                return (100 * temp_code[1]) + temp_code[2]
            # END IF
        # END INNER FOR
    # END FOR

    return -1
# END get_solution()


if __name__ == '__main__':
    main()
# END IF
